sticky-rows/StickyRows.js

```diff
@@ -0,0 +1,279 @@
+        && toggle !== undefined;
+
+    // setup config for this sheet in the spreadsheet attributes
+    const key = "stickyRows." + sheet.getSheetId();
+    const config = mergeConfig_(key, {
+        dynamic_range: dynamic_range,
+        sticky_range: sticky_range,
+        controlCell: argsValid ? activeRange.getA1Notation() : ''
+    });
+
+    if (!argsValid) {
+        // We must always update config.controlCell first, to ensure onChangeStickyRows() doesn't overwrite
+        // what was previously a control cell with a formula but may have been modified.
+        throw new Error("Please specify all three arguments: dynamic_range (string), sticky_range (string) and some random integer as third argument");
+    }
+
+    // ensure onChange() trigger
+    let haveTrigger = false;
+    for (const trigger of ScriptApp.getProjectTriggers()) {
+        if (trigger.getEventType() === ScriptApp.EventType.ON_CHANGE && trigger.getHandlerFunction() === TRIGGER_FUNCTION_NAME) {
+            haveTrigger = true;
+        }
+    }
+
+    if (!haveTrigger) {
+        throw new Error(`Please install function ${TRIGGER_FUNCTION_NAME}() as trigger in this Spreadsheet's script context (event source: From spreadsheet, event type: onChange)`)
+    }
+
+    if (config.error) {
+        throw new Error(config.error);
+    }
+}
+
+
+/** Uninstall triggers. */
+function uninstall() {
+    TriggerUtil.uninstall(TRIGGER_HANDLER_FUNCTION);
+}
+
+
+/** Install spreadsheet onChange trigger. */
+function install() {
+
+    TriggerUtil.uninstall(TRIGGER_HANDLER_FUNCTION);
+
+    Logger.log("Installing onChange trigger for spreadsheet %s", spreadsheet.getId());
+
+    const sheet = SpreadsheetApp.getActive();
+    ScriptApp.newTrigger(TRIGGER_HANDLER_FUNCTION)
+        .forSpreadsheet(SpreadsheetApp.getActiveSpreadsheet().getId())
+        .onChange()
+        .create();
+    Logger.log("Installed onChange trigger for spreadsheet %s", sheet.getSpreadsheet());
+}
+
+
+/** Perform the actual sticky rows logic.
+ *
+ * Supposed to be called in a SpreadSheet.onChange() trigger.
+ */
+function onChangeStickyRows(event) {
+
+    const spreadsheet = event.source || SpreadsheetApp.getActiveSpreadsheet();
+    if (!spreadsheet) {
+        throw new Error("Must run in the context of a Spreadsheet");
+    }
+
+    let sheet;
+    try {
+        sheet = spreadsheet.getActiveRange().getSheet();
+    } catch (e) {
+        sheet = spreadsheet.getActiveSheet();
+    }
+
+    const key = "stickyRows." + sheet.getSheetId();
+    const config = mergeConfig_(key, {});
+    if (!config.dynamic_range || !config.sticky_range || !config.controlCell) {
+        // STICKY_ROWS() also reports this error, we can't forward since we know no control cell here
+        throw new Error("Please call STICK_ROWS() with a valid dynamic_range and sticky_range in a formula on the sheet");
+    }
+
+    try {
+        realignStickyRows_(sheet, config);
+        if (config.error) {
```

Comment:
What actually writes config.error? Seems magical to me where this is populated

Reply:
Yes, `mergeConfig_()` has a side-effect (read + de-serialize, merge, serialize + write the DocumentProperty). 🙄